src/java.base/share/classes/java/security/PEMDecoder.java

```diff
@@ -0,0 +1,427 @@
+    /**
+     * Returns an instance of {@code PEMDecoder}.
+     *
+     * @return returns a {@code PEMDecoder}
+     */
+    public static PEMDecoder of() {
+        return PEM_DECODER;
+    }
+
+    /**
+     * After the header, footer, and base64 have been separated, identify the
+     * header and footer and proceed with decoding the base64 for the
+     * appropriate type.
+     */
+    private DEREncodable decode(PEMRecord pem) {
+        Base64.Decoder decoder = Base64.getMimeDecoder();
+
+        try {
+            return switch (pem.type()) {
+                case Pem.PUBLIC_KEY -> {
+                    X509EncodedKeySpec spec =
+                        new X509EncodedKeySpec(decoder.decode(pem.pem()));
+                    yield (getKeyFactory(spec.getAlgorithm())).
+                        generatePublic(spec);
+                }
+                case Pem.PRIVATE_KEY -> {
+                    PKCS8Key p8key = new PKCS8Key(decoder.decode(pem.pem()));
+                    String algo = p8key.getAlgorithm();
+                    KeyFactory kf = getKeyFactory(algo);
+                    DEREncodable d = kf.generatePrivate(
+                        new PKCS8EncodedKeySpec(p8key.getEncoded(), algo));
+
+                    // Look for a public key inside the pkcs8 encoding.
+                    if (p8key.getPubKeyEncoded() != null) {
+                        // Check if this is a OneAsymmetricKey encoding
+                        X509EncodedKeySpec spec = new X509EncodedKeySpec(
+                            p8key.getPubKeyEncoded(), algo);
+                        yield new KeyPair(getKeyFactory(algo).
+                            generatePublic(spec), (PrivateKey) d);
+
+                    } else if (d instanceof PKCS8Key p8 &&
+                        p8.getPubKeyEncoded() != null) {
+                        // If the KeyFactory decoded an algorithm-specific
+                        // encodings, look for the public key again.  This
+                        // happens with EC and SEC1-v2 encoding
+                        X509EncodedKeySpec spec = new X509EncodedKeySpec(
+                            p8.getPubKeyEncoded(), algo);
+                        yield new KeyPair(getKeyFactory(algo).
+                            generatePublic(spec), p8);
+                    } else {
+                        // No public key, return the private key.
+                        yield d;
+                    }
+                }
+                case Pem.ENCRYPTED_PRIVATE_KEY -> {
+                    if (password == null) {
+                        yield new EncryptedPrivateKeyInfo(decoder.decode(
+                            pem.pem()));
+                    }
+                    yield new EncryptedPrivateKeyInfo(decoder.decode(pem.pem())).
+                        getKey(password.getPassword());
+                }
+                case Pem.CERTIFICATE, Pem.X509_CERTIFICATE -> {
+                    CertificateFactory cf = getCertFactory("X509");
```

Comment:
Use "X.509". "X509" is an alias and may not be supported by other JDK implementations. Same comment on line 196.

Reply:
This issue still needs to be fixed. Change `getCertFactory("X509")` to `getCertFactory("X.509")`